### PR TITLE
Cell maps ceil rules

### PR DESCRIPTION
**Pull request by traiansf**

### CODE REVIEW

File: k-distribution/include/builtin/domains.k
Comment (ttuegel) [inline k-distribution/include/builtin/domains.k:205]: Why are we retaining these rules for the Haskell backend?

File: kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java
Comment (ttuegel) [inline kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java]: What's going on with the pairs of set and element variables?

File: kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java
Comment (ttuegel) [inline kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java]: Should the `#Ceil` and `#And` constants be defined somewhere?

File: kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java
Comment (ttuegel) [inline kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java:598]: What is the `anywhere` for?

File: k-distribution/include/builtin/domains.k
Comment (traiansf) [reply k-distribution/include/builtin/domains.k:205]: Well, I figured that since the Java engine still needs to work, we must keep these around.
One way to exclude them from the Haskell backend all-together would be to put them into a module which has attributes `symbolic` and `kast`, which would make it acceptable to the Java backend but excluded from the Haskell backend.  I'll try and do that.

File: kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java
Comment (traiansf) [reply kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java]: the element variables are used for the in_keys rules. The set variables are used for the ceil rules.

File: kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java
Comment (traiansf) [reply kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java:598]: That's awkward, but since #Ceil is not a function, we usually need to annotate `#Ceil` rules with `anywhere` to prevent them from being completed with `<k>` and stuff. Maybe here I didn't need to add them, as we're late in the compilation process, but I added them for consistency.

File: k-distribution/include/builtin/domains.k
Comment (traiansf) [reply k-distribution/include/builtin/domains.k:205]: I've completely separated the existing symbolic part from the Haskell part. We can decide later whether we want to share some of the rules or not.

**Review by dwightguth — APPROVED** [review]

File: k-distribution/include/builtin/domains.k
Comment (ttuegel) [inline k-distribution/include/builtin/domains.k:177]: Why isn't `_[_ <- _]` simply `functional`?

File: kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java
Comment (ttuegel) [reply kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java]: I find it difficult to follow how the rules are constructed because the construction of the rules is interleaved in this way, but it's not difficult enough that I'm going to hold up this pull request.

**Review by ttuegel — APPROVED** [review]

File: k-distribution/include/builtin/domains.k
Comment (traiansf) [reply k-distribution/include/builtin/domains.k:177]: You're right. `_[_ <- _]` is already functional. the rule is not necessary.

**Comment by traiansf** [comment]
@dwightguth : this is the error I'm getting from `llvm-kompile`:
```
/home/traian/Documents/k/k-distribution/lib/../bin/llvm-kompile: line 27: 29891 Segmentation fault      (core dumped) "$(dirname "$0")"/llvm-kompile-codegen "$definition" "$dt_dir"/dt.yaml "$dt_dir" $debug > "$mod"
```

I tried to run `llvm-kompile-codegen` stand-alone, but I get the same Segmentation fault error, and I don't know where to find the core dump.
```
/home/traian/Documents/k/k-distribution/bin/llvm-kompile-codegen imp-kompiled/definition.kore imp-kompiled/dt/dt.yaml imp-kompiled/dt 1
Segmentation fault (core dumped)
```

**Comment by traiansf** [comment]
Given the above, I'm going to merge this as-is (once the tests pass), but if you can take a look and help me tackle the llvm backend issue, I'd like to remove the haskell backend guard in a subsequent PR.

File: kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java
Comment (traiansf) [reply kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java]: I've refactored the code and added comments with the rules. hope it is clearer now.

**Pull request merged**
